src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/WireMessage.java

```diff
@@ -120,37 +69,127 @@ public static WireMessage create(ConnectionHandler connection, DataInputStream i
 
   /**
    * Once the message is ready, call this to send it across the wire.
+   * Effectively a template pattern.
    *
    * @throws Exception If the sending fails.
    */
-  public abstract void send() throws Exception;
+  public void send() throws Exception {
+    logger.log(Level.FINE, this.toString());
+    sendPayload();
+  }
+
+  /**
+   * Override this method to include post-processing and metadata in the sending process. Template
+   * method for send.
+   *
+   * @throws Exception If any step in output message fails.
+   */
+  protected abstract void sendPayload() throws Exception;
+
+  /**
+   * Used for logging.
+   *
+   * @return The official name of the wire message.
+   */
+  protected abstract String getMessageName();
+
+  /**
+   * Used for logging.
+   *
+   * @return Payload metadata.
+   */
+  protected abstract String getPayloadString();
+
+  /**
+   * Used for logging.
+   *
+   * @return Message Identifier (int for Bootstrap, char otherwise).
+   */
+  protected abstract String getIdentifier();
+
+  @Override
+  public String toString() {
+    return new MessageFormat("> Received Message: ({0}) {1}, with Payload: '{'{2}'}'")
+        .format(new Object[]{
+            this.getIdentifier(),
+            this.getMessageName(),
+            this.getPayloadString()
+        });
+  }
 
   /**
-   * Read the remainder of the received message. To be called once metdata for the message is read.
+   * Read the remainder of the received message. To be called once metadata for the message is read.
    * Metadata is designated as type, length, etc.
    *
-   * @param input The input stream containing the remainder of the message.
    * @return The remainder of the message in String format.
    * @throws Exception If reading fails in any way.
    */
-  protected String read(DataInputStream input) throws Exception {
-    return PGWireProtocol.readString(input, this.length - this.remainder);
+  protected String readAll() throws Exception {
+    return read(this.length - this.getHeaderLength());
   }
 
+  /**
+   * Reads a fixed-length string from teh saved stream. The string still needs to be
+   * null-terminated, but the read is more efficient, as we can read the entire string in one go
+   * instead of continuously checking for a null-terminator.
+   *
+   * @param length The number of bytes to read.
+   * @return the string.
+   * @throws IOException if an error occurs while reading from the stream, or if no null-terminator
+   * is found at the end of the string.
+   */
+  public String read(int length) throws IOException {
+    byte[] buffer = new byte[length - 1];
+    this.inputStream.readFully(buffer);
```

Comment:
I would need to read the buffer either way to get the data that comes after, so it wouldn't really gain. Making one single call might save a microsecond due to branch prediction, but then we would need to access it either way to ensure it is null terminated and formatted properly, so I'm unsure how much it would save.